Add a Primary Person List Collector block.

bitdivision opened the pull request:
### What is the context of this PR?
Allows collecting one person in a list who is stored as a primary person in the list store.

Designed as per https://github.com/ONSdigital/eq-survey-runner/pull/2126/

I have tried to keep functionality somewhat seperated from core list collector functionality since it is likely to be pulled out after census.

The functionality in perform_list_action is becoming unwieldy. I'm looking at how to refactor that now.

### How to review 
Compare the journey in test_list_collector_primary_person to the journey in v13 prototype of WLH.


### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

MebinAbraham commented on app/data_model/list_store.py:
```diff
@@ -12,24 +13,80 @@ def random_string(length):
+            context = dict()
+
+        self.list_items = list_items
+        self.context = context
```
Instead of the ifs?
```python
class ListRepresentation:
    def __init__(self, list_items: List = None, context: Mapping = None):
        self.list_items = list_items or []
        self.context = context or {}

```

bitdivision replied:
Fixed.

bambrp commented on templates/primarypersonlistaddoreditquestion.html:
```diff
@@ -0,0 +1 @@
+{% extends 'question.html' %}
```
If not overriding anything, should this be an include too, in common with primarypersonlistcollector.html?

MebinAbraham commented on app/data_model/list_store.py:
```diff
@@ -12,24 +13,80 @@ def random_string(length):
+        return self.list_items == other.list_items and self.context == other.context
+
+    def remove_item_id(self, to_remove):
+        self.list_items.remove(to_remove)
```
Currently, if I pass the primary persons list item id to the remove URL, it removes them. And then results in a 500 if I change the answer to `No` since the id no longer exists.
```sh
File "eq-survey-runner/app/data_model/list_store.py", line 33, in remove_item_id
    self.list_items.remove(to_remove)
ValueError: list.remove(x): x not in list
```

bitdivision replied:
Fixed.

MebinAbraham commented on app/data_model/list_store.py:
```diff
@@ -12,24 +13,80 @@ def random_string(length):
+            x.get('name'): ListRepresentation(
+                list_items=x.get('items'), context=x.get('context', {})
+            )
+            for x in existing_items
```
`for existing_item in existing_items`? or just refer to them as list_items?

bitdivision replied:
I'm using `item` now, keeps it a little shorter

bitdivision replied:
Fixed.

bambrp commented on tests/functional/generate_pages.py:
```diff
@@ -194,11 +194,11 @@
 """
 
-LIST_SUMMARY_EDIT_LINK_GETTER = r"""  listEditLink(instance) { return `tbody:nth-child(${instance}) td:last-child a:first-child`; }
+LIST_SUMMARY_EDIT_LINK_GETTER = r"""  listEditLink(instance) { return `tbody:nth-child(${instance}) td:last-child a[aria-label^='Change'`; }
```
This works great for now, though I would be happier still with something that wasn't subject to a potential content change (i.e. "could you make this say 'alter' instead?") Perhaps a template change to add a class we could select on, or a non user-visible attribute value of our own invention, such as eq-primary-person?

bitdivision replied:
I think the only other reasonable way to do it would be to add a data tag. We do this elsewhere, but this would require a change to the pattern library in this case I think.

bambrp replied:
I hadn't known of the data-* attributes before, that would be a good way to do this, I like the side effects of populating the element.dataset in the DOM. Handy.

bitdivision replied:
Fixed with data attribute

MebinAbraham reviewed (changes requested):
POST to `/questionnaire/people/ZlwafC/add-or-edit-primary-person/` is currently not idempotent.
<img width="699" alt="Screenshot 2019-07-04 at 12 52 04" src="https://user-images.githubusercontent.com/35296336/60664955-1ac76b00-9e5b-11e9-92e9-6fa2c15f28f5.png">

bitdivision commented:
Duplicates are now fixed @MebinAbraham

ajmaddaford commented on app/data_model/list_store.py:
```diff
@@ -48,42 +100,64 @@ def is_dirty(self):
+            del self[list_name].context['primary_person']
+
+    def get_primary_person(self, list_name):
+        return self[list_name].context.get('primary_person')
```
This is specific to primary person. It would be cleaner if this was either `get_context(list_name)` and the caller extracts what it needs from the returned context, or `get_context_item(list_name, context_key)`.

bitdivision replied:
As discussed, changed this class to use a `primary_person` class property. 

Though this does pollute the list_store with a primary person, I don't think any further functionality will be built on top of this and it should be simple to rip out if we want to.

ajmaddaford commented on app/data_model/list_store.py:
```diff
@@ -48,42 +103,64 @@ def is_dirty(self):
+        return self[list_name].context.get('primary_person')
+
+    def add_list_item(
+        self, list_name, list_item_id: str = None, context: Mapping = None
```
Why has this been changed to allow a predetermined list id?

bitdivision replied:
Fixed.

ajmaddaford commented on app/data_model/list_store.py:
```diff
@@ -48,42 +103,64 @@ def is_dirty(self):
 
-        named_list.append(new_list_identifier)
+        if context:
+            named_list.context = context
```
I think the context needs to be separated out from the list items. Currently the full context is overwritten on adding a single list item. 

Also, we should consider making context a string for the caller as it's unnecessary to provide the list item id twice e.g. `add_context(list_item_id=123456, context='primary-person')`

bitdivision replied:
Changed.

ajmaddaford commented on app/data_model/list_store.py:
```diff
@@ -12,24 +13,78 @@ def random_string(length):
     return ''.join(random.choice(ascii_letters) for _ in range(length))
 
 
+class ListRepresentation:
```
Possibly just `List`?

bitdivision replied:
This one is annoying because List conflicts with `typing.List`. I've been struggling to think of anything better.

iwootten replied:
ListModel?

bitdivision replied:
Sounds reasonable to me. Any objections @ajmaddaford ?

ajmaddaford replied:
👍 It's better than anything I can think of!

JamesGardiner replied:
ListyMcListFace

ajmaddaford commented on app/data_model/list_store.py:
````diff
@@ -12,24 +13,78 @@ def random_string(length):
+    ```
+    self._lists: {
+        list_name_1: ListRepresentation(
+            list_items: [1,2,3],
````
`list_items` could be shortened to `items`.

bitdivision replied:
Done

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -514,32 +514,40 @@ def map_list_collector_config(
+            )
+
+        if list_item.get('primary_person'):
+            rows.insert(0, new_row)
```
This shouldn't be needed as the list items should be in the correct order (they have to be so we can iterate over the sections in the right order).

bitdivision replied:
Fixed.

ajmaddaford commented on app/questionnaire/questionnaire_store_updater.py:
```diff
@@ -38,13 +38,34 @@ def remove_answers(self, answer_ids: List):
+
+        self._list_store.add_list_item(
+            list_name, list_item_id=list_item_id, context=context
+        )
```
I think it would be much clearer to have an `add_context` method.

bitdivision replied:
Probably not relevant anymore with `primary_person` property/.

ajmaddaford commented on app/templating/view_context.py:
```diff
@@ -92,24 +95,32 @@ def build_view_context_for_list_collector(
         answer['id'] for answer in rendered_block['add_block']['question']['answers']
     ]
 
+    primary_person = list_store.get_primary_person(list_name)
```
We need to review the use of context for the primary person. The list store abstracts it into a context, but here we're adding primary person specific logic to the view context.

MebinAbraham reviewed (commented):
As discussed in IRL @bitdivision, at the moment since the list item id is not stored in the list store before visiting the primary person page, you are able to POST arbitrary ids and essentially change primary person's id with an existing list item id etc.

So as planned, we should generate and store the list item id for the primary person before visiting the primary person add or edit block.

JamesGardiner commented on app/data_model/list_store.py:
```diff
@@ -12,22 +13,92 @@ def random_string(length):
+            self.primary_person = None
+
+    @classmethod
+    def from_dict(cls, input_dict: dict):
```
Should probably use the `typing.Mapping` type here to be more generic in acceptance

JamesGardiner commented on app/data_model/list_store.py:
```diff
@@ -12,22 +13,92 @@ def random_string(length):
-    def __init__(self, existing_items=None):
-        if not existing_items:
-            existing_items = []
+    def __init__(self, existing_items: Mapping[str, ListRepresentation] = None):
```
Is this type hint accurate? I think it should be `Union[Mapping[str, ListRepresentation], None]`

JamesGardiner commented on app/data_model/list_store.py:
```diff
@@ -48,42 +119,69 @@ def is_dirty(self):
-        ]
+        output = []
+
+        for name, list_representation in self._lists.items():
```
Could keep this as a list comp:
```Python
return [list_representaion.to_dict(name=name) for name, list_representaion in self._lists.items()]
```

JamesGardiner commented on app/views/questionnaire.py:
```diff
@@ -326,12 +333,14 @@ def post_block_handler(
 
 
-def perform_list_action(
+def perform_list_action(  # noqa: C901 pylint: disable=too-complex
```
Can this not be refactored rather than disabling lint warnings?

bitdivision replied:
Ideally yes. I'm still looking at how best to do that.

JamesGardiner commented on scripts/test_schemas.sh:
```diff
@@ -6,8 +6,8 @@ if [ "$1" == "--local" ] || [ "$2" == "--local" ]; then
 if [ "$run_docker" == true ]; then
-    docker pull onsdigital/eq-schema-validator:v3
-    validator="$(docker run -d -p 5001:5000 onsdigital/eq-schema-validator:v3)"
+    docker pull onsdigital/eq-schema-validator:3092-implement-primary-person-list
```
Reminder to revert these before merge

JamesGardiner commented on tests/app/data_model/test_list_store.py:
```diff
@@ -1,18 +1,19 @@
 from unittest.mock import patch
-from app.data_model.list_store import ListStore
+import pytest
```
Although this is a test module, it should still follow PEP8 standards for import ordering. Do these not get parsed by the linter?

bitdivision replied:
This one is still in-progress...

bitdivision replied:
It does still pass linting though? I can't see anything wrong there, std library > external libraries > app imports?

JamesGardiner replied:
I was under the impression that a blank line needed to be separating them to pass linting?

e.g.
```Python
from unittest.mock import patch

import pytest
```

JamesGardiner replied:
As per PEP8:

```
Imports should be grouped in the following order:

    Standard library imports.
    Related third party imports.
    Local application/library specific imports.

You should put a blank line between each group of imports.
```

bitdivision replied:
Ah, I don't think that's enforced. I'll add it in though.

bitdivision commented:
Going to wait until #2155 goes in before making any further changes to questionnaire.py on here.

iwootten commented on app/views/questionnaire.py:
```diff
@@ -465,6 +483,68 @@ def perform_list_action(
     return list_collector_url
 
 
+def handle_primary_person_action(
```
There are a lot of args to this method. I think a couple (router/routing_path) could be removed/replaced if rather than return the next url it was determined following the questionnaire store update?

bitdivision replied:
The problem with this is that you end up having a load of if statements for the questionnaire store update, and then a load more ifs when you come to work out the next URL, since each of these block types have slightly different behaviour for both of those.

ajmaddaford commented on app/submitter/convert_payload_0_0_3.py:
```diff
@@ -33,40 +33,57 @@ def convert_answers_to_payload_0_0_3(
+    }
+    if collector in sub_block_map.keys():
+        sub_block_name = sub_block_map[collector]
+        answers_in_add_block = get_answers_for_list_collector_sub_block(
```
I think it would be more appropriate to add a method on `QuestionnaireSchema` to get the answer ids for list items, something like `get_answer_ids_for_list_items`. This would keep some block specific logic out of this method.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -565,16 +566,24 @@ def perform_list_action(
+    if block['type'] == 'ListEditQuestion':
         questionnaire_store_updater.update_answers(form)
 
+    if schema.is_primary_person_block_type(block['type']):
```
I can't see why primary person uses the `perform_list_action` method. It would be clearer if this was called from `post_block_handler`.

bitdivision replied:
Done. Though annoyingly makes post_handler 'too complex'. I've ignored this for now on the basis that this is all going to be refactored fairly soon.

MebinAbraham commented on app/data_model/list_store.py:
```diff
@@ -12,22 +13,91 @@ def random_string(length):
+        return serialised
+
+    def __repr__(self):
+        return f'<ListRepresentation name={self.name} items={self.items}, primary_person={self.primary_person}>'
```
```suggestion
        return f'<ListModel name={self.name} items={self.items}, primary_person={self.primary_person}>'
```

MebinAbraham commented on app/data_model/list_store.py:
````diff
@@ -12,22 +13,91 @@ def random_string(length):
+
+    ```
+    self._lists: {
+        list_name_1: ListRepresentation(
````
```suggestion
        list_name_1: ListModel(
```

MebinAbraham reviewed (commented):
Looks good.

Is the previous link from a list collector supposed to take you to the driving question or should it go to the edit screen?. 

Or should there not be a previous link?. Non-primary person list collector doesn't have the previous link, not sure if that was intentional.

MebinAbraham reviewed (changes requested):
So there is no remove link for the primary person but if I manually POST to the remove person endpoint with the list item id of the primary person then I am able to remove the primary person. We should prevent this?

1. Add a primary person.
2. Add another person.
3. Click remove on 'another' person.
4. Change list item id in url to the primary person's list item id
5. Click Continue.

bitdivision commented:
The previous link is meant to take you back to the driving question, not the edit page.

https://github.com/ONSdigital/eq-survey-runner/blob/56932bc2302dc54755a1276881cb8bb507ed72ae/doc/architecture/decisions/0009-primary-person.md

You have just reminded me that I need to update that design doc though :)

bitdivision commented:
> So there is no remove link for the primary person but if I manually POST to the remove person endpoint with the list item id of the primary person then I am able to remove the primary person. We should prevent this?
> 
>     1. Add a primary person.
> 
>     2. Add another person.
> 
>     3. Click remove on 'another' person.
> 
>     4. Change list item id in url to the primary person's list item id
> 
>     5. Click Continue.

I think this came up before and the conclusion was that it was possible to remove a primary person by answering 'No', so this scenario wasn't a huge issue. 

Thinking about it more though, downstream would receive an inconsistent payload. The primary-person answer would be 'Yes' but that would not be reflected in the list store.

I'll see how much extra logic this would introduce, but would appreciate thoughts @ajmaddaford @JamesGardiner @MebinAbraham

bitdivision commented:
I am going to make the change to disallow removing a primary person from the list collector. The change is simple, and should avoid the possibility of some strange routing bugs.

bitdivision commented:
@ajmaddaford @JamesGardiner @MebinAbraham I've made that change now.

MebinAbraham reviewed (approved):
👍Thailand awaits.

bitdivision commented:
Will wait for @ajmaddaford before squashing.

bitdivision committed:
Add Primary Person List Collector

Changes the list store representation and includes a primary person
attribute on the ListStore

Includes changes to the list summary to call out the primary person
